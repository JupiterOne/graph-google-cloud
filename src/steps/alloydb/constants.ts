export const STEP_ALLOYDB_POSTGRE_SQL_SERVICE = 'create-postgre-sql-service';
export const STEP_ALLOYDB_POSTGRE_SQL_CLUSTER =
  'fetch-alloydb-postgre-sql-cluster';
export const STEP_ALLOYDB_POSTGRE_SQL_INSTANCE =
  'fetch-alloydb-postgre-sql-instance';
export const STEP_ALLOYDB_POSTGRE_SQL_CONNECTION =
  'fetch-alloydb-postgre-sql-connection';
export const STEP_ALLOYDB_POSTGRE_SQL_BACKUP =
  'fetch-alloydb-postgre-sql-backup';

// Entity type
export const ENTITY_TYPE_ALLOYDB_POSTGRE_SQL_SERVICE = 'google_cloud_alloydb';
export const ENTITY_CLASS_ALLOYDB_POSTGRE_SQL_SERVICE = ['Service'];

export const ENTITY_TYPE_POSTGRE_SQL_ALLOYDB_CLUSTER =
  'google_cloud_alloydb_cluster';
export const ENTITY_CLASS_POSTGRE_SQL_CLUSTER = [
  'Database',
  'DataStore',
  'Cluster',
];

export const ENTITY_TYPE_POSTGRE_SQL_ALLOYDB_INSTANCE =
  'google_cloud_alloydb_instance';
export const ENTITY_CLASS_POSTGRE_SQL_INSTANCE = [
  'Database',
  'DataStore',
  'Host',
];

export const ENTITY_TYPE_POSTGRE_SQL_CONNECTION =
  'google_cloud_alloydb_connection';
export const ENTITY_CLASS_POSTGRE_SQL_CONNECTION = ['Network'];

export const ENTITY_TYPE_POSTGRE_SQL_BACKUP = 'google_cloud_alloydb_backup';
export const ENTITY_CLASS_POSTGRE_SQL_BACKUP = ['Backup'];

// Relationships
export const STEP_PROJECT_HAS_ALLOYDB_SERVICE_RELATIONSHIP =
  'build-project-alloydb-service-relationship';
export const STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP =
  'build-project-has-alloydb-cluster-relationship';

export const STEP_ALLOYDB_INSTANCE_USES_CLUSTER_RELATIONSHIP =
  'build-alloydb-instance-uses-cluster-relationship';

export const STEP_ALLOYDB_CLUSTER_HAS_BACKUP_RELATIONSHIP =
  'build-alloydb-cluster-has-backup-relationship';

export const STEP_ALLOYDB_INSTANCE_HAS_CONNECTION_RELATIONSHIP =
  'build-alloydb-instance-has-connection-relationship';

export const STEP_ALLOYDB_CLUSTER_USES_KMS_KEY_RELATIONSHIP =
  'build-alloydb-cluster-uses-kms-key-relationship';

export const STEP_USER_ASSIGNED_ALLOYDB_CLUSTER_RELATIONSHIP =
  'build-user-asssigned-alloydb-cluster-relatipnship';

export const RELATIONSHIP_TYPE_PROJECT_HAS_ALLOYDB_SERVICE =
  'google_cloud_project_has_alloydb';
export const RELATIONSHIP_TYPE_PROJECT_HAS_ALLOYDB_CLUSTER =
  'google_cloud_project_has_alloydb_cluster';
export const RELATIONSHIP_TYPE_ALLOYDB_INSTANCE_USES_CLUSTER =
  'google_cloud_alloydb_instance_uses_cluster';
export const RELATIONSHIP_TYPE_ALLOYDB_CLUSTER_HAS_BACKUP =
  'google_cloud_alloydb_cluster_has_backup';
export const RELATIONSHIP_TYPE_STEP_ALLOYDB_INSTANCE_HAS_CONNECTION_BACKUP =
  'google_cloud_alloydb_instance_has_connection';
export const RELATIONSHIP_TYPE_ALLOYDB_CLUSTER_USES_KMS_KEY =
  'google_cloud_alloydb_cluster_uses_kms_crypto_key';
export const RELATIONSHIP_TYPE_USER_ASSIGNED_ALLOYDB_CLUSTER =
  'google_user_assigned_cloud_alloydb_cluster';

// IngestionSources
export const IngestionSources = {
  ALLOYDB_POSTGRE_SQL_SERVICE: 'postgre-sql-service',
  ALLOYDB_POSTGRE_SQL_CLUSTER: 'alloydb-postgre-sql-cluster',
  ALLOYDB_POSTGRE_SQL_INSTANCE: 'alloydb-postgre-sql-instance',
  ALLOYDB_POSTGRE_SQL_CONNECTION: 'alloydb-postgre-sql-connection',
  ALLOYDB_POSTGRE_SQL_BACKUP: 'alloydb-postgre-sql-backup',
  PROJECT_HAS_ALLOYDB_SERVICE_RELATIONSHIP:
    'project-has-alloydb-service-relationship',
  PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP:
    'project-has-alloydb-cluster-relationship',
  ALLOYDB_INSTANCE_USES_CLUSTER_RELATIONSHIP:
    'alloydb-instance-uses-cluster-relationship',
  ALLOYDB_CLUSTER_HAS_BACKUP_RELATIONSHIP:
    'alloydb-cluster-has-backup-relationship',
  ALLOYDB_INSTANCE_HAS_CONNECTION_RELATIONSHIP:
    'alloydb-instance-has-connection-relationship',
  ALLOYDB_CLUSTER_USES_KMS_KEY_RELATIONSHIP:
    'alloydb-cluster-uses-kms-key-relationship',
  GOOGLE_USER_ASSIGNED_ALLOYDB_CLUSTER_RELATIONSHIP:
    'google-user-assigned-alloydb-cluster-relationship',
};

// IngestionSources Configs
export const AlloyDBIngestionConfig = {
  [IngestionSources.ALLOYDB_POSTGRE_SQL_SERVICE]: {
    title: 'AlloyDB for Postgre SQL service',
    description: 'AlloyDB for Postgre SQL service.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_POSTGRE_SQL_CLUSTER]: {
    title: 'AlloyDB for Postgre SQL Cluster',
    description: 'AlloyDB for Postgre SQL Cluster.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_POSTGRE_SQL_INSTANCE]: {
    title: 'AlloyDB for Postgre SQL Instance',
    description: 'AlloyDB for Postgre SQL Instance.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_POSTGRE_SQL_CONNECTION]: {
    title: 'AlloyDB for Postgre SQL Connection',
    description: 'AlloyDB for Postgre SQL Connection.',
    defaultsToDisabled: false,
  },
  [IngestionSources.PROJECT_HAS_ALLOYDB_SERVICE_RELATIONSHIP]: {
    title: 'Project AlloyDB Service Has Relationship',
    description: 'Build Project AlloyDB Service Has Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP]: {
    title: 'Project AlloyDB Cluster Has Relationship',
    description: 'Build Project AlloyDB Cluster Has Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_INSTANCE_USES_CLUSTER_RELATIONSHIP]: {
    title: 'AlloyDB Instance Cluster Uses Relationship',
    description: 'Build AlloyDB Instance Cluster Uses Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_CLUSTER_HAS_BACKUP_RELATIONSHIP]: {
    title: 'AlloyDB Instance Backup Has Relationship',
    description: 'Build AlloyDB Backup Cluster Has  Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_INSTANCE_HAS_CONNECTION_RELATIONSHIP]: {
    title: 'AlloyDB Instance Connection Has  Relationship',
    description: 'Build AlloyDB Instance Connection Has Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_CLUSTER_USES_KMS_KEY_RELATIONSHIP]: {
    title: 'AlloyDB Cluster Kms Key Uses Relationship',
    description: 'Build AlloyDB Cluster Kms Key Has Relationship.',
    defaultsToDisabled: false,
  },
};

// IAM Permissions
export const AlloyDBPermissions = {
  STEP_ALLOYDB_POSTGRE_SQL_CLUSTER: ['alloydb.clusters.list'],
  STEP_ALLOYDB_POSTGRE_SQL_INSTANCE: ['alloydb.instances.list'],
  STEP_ALLOYDB_POSTGRE_SQL_CONNECTION: ['alloydb.instances.connect'],
  STEP_ALLOYDB_POSTGRE_SQL_BACKUP: ['alloydb.backups.get'],
};
