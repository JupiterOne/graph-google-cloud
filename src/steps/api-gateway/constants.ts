export const STEP_API_GATEWAY_APIS = 'fetch-api-gateway-apis';
export const STEP_API_GATEWAY_API_CONFIGS = 'fetch-api-gateway-api-configs';
export const STEP_API_GATEWAY_GATEWAYS = 'fetch-api-gateway-gateways';

export const ENTITY_CLASS_API_GATEWAY_API = 'Service';
export const ENTITY_TYPE_API_GATEWAY_API = 'google_api_gateway_api';

export const ENTITY_CLASS_API_GATEWAY_API_CONFIG = 'Configuration';
export const ENTITY_TYPE_API_GATEWAY_API_CONFIG =
  'google_api_gateway_api_config';

export const ENTITY_CLASS_API_GATEWAY_GATEWAY = 'Gateway';
export const ENTITY_TYPE_API_GATEWAY_GATEWAY = 'google_api_gateway_gateway';

export const RELATIONSHIP_TYPE_API_GATEWAY_API_USES_CONFIG =
  'google_api_gateway_api_uses_config';
export const RELATIONSHIP_TYPE_API_GATEWAY_API_HAS_GATEWAY =
  'google_api_gateway_api_has_gateway';
export const RELATIONSHIP_TYPE_API_GATEWAY_API_CONFIG_USES_SERVICE_ACCOUNT =
  'google_api_gateway_api_config_uses_iam_service_account';

export const IngestionSources = {
  API_GATEWAY_APIS: 'api-gateway-apis',
  API_GATEWAY_API_CONFIGS: 'api-gateway-api-configs',
  API_GATEWAY_GATEWAYS: 'api-gateway-gateways',
};

export const ApiGatewayIngestionConfig = {
  [IngestionSources.API_GATEWAY_APIS]: {
    title: 'Google Cloud API Gateway APIs',
    description: 'Endpoint management for API gateways.',
    defaultsToDisabled: false,
  },
  [IngestionSources.API_GATEWAY_API_CONFIGS]: {
    title: 'Google Cloud API Gateway API Configurations',
    description: 'Config settings for API interfaces.',
    defaultsToDisabled: false,
  },
  [IngestionSources.API_GATEWAY_GATEWAYS]: {
    title: 'Google Cloud API Gateway Gateways',
    description: 'Networking gateways for API management.',
    defaultsToDisabled: false,
  },
};

export const ApiGatewayPermissions = {
  STEP_API_GATEWAY_APIS: [
    'apigateway.apis.getIamPolicy',
    'apigateway.apis.list',
  ],
  STEP_API_GATEWAY_API_CONFIGS: [
    'apigateway.apiconfigs.list',
    'apigateway.apiconfigs.getIamPolicy',
  ],
  STEP_API_GATEWAY_GATEWAYS: [
    'apigateway.gateways.list',
    'apigateway.gateways.getIamPolicy',
  ],
};
