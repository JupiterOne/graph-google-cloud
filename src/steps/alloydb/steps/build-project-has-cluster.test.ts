import {
  executeStepWithDependencies,
  Recording,
  StepTestConfig,
} from '@jupiterone/integration-sdk-testing';
import { invocationConfig } from '../../..';
import { integrationConfig } from '../../../../test/config';
import {
  getMatchRequestsBy,
  setupGoogleCloudRecording,
} from '../../../../test/recording';
import { STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP } from '../constants';

describe(STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(450000);

  const integrationConfigNew = {
    ...integrationConfig,
    serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
      'j1-gc-integration-dev-v2',
      'j1-gc-integration-dev-v3',
    ),
    serviceAccountKeyConfig: {
      ...integrationConfig.serviceAccountKeyConfig,
      project_id: 'j1-gc-integration-dev-v3',
    },
  }

  test(STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfigNew),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP,
      instanceConfig: integrationConfigNew,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});
