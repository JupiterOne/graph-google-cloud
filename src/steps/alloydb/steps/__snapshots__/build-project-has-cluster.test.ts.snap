// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Project has AlloyDB Cluster Relationship should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Database",
        "DataStore",
        "Cluster",
      ],
      "_key": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/asia-south1/clusters/test-cluster-03",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "automatedBackupPolicy": {
              "backupWindow": "3600s",
              "enabled": false,
              "location": "asia-south1",
              "timeBasedRetention": {
                "retentionPeriod": "1209600s",
              },
              "weeklySchedule": {
                "daysOfWeek": [
                  "MONDAY",
                  "TUESDAY",
                  "WEDNESDAY",
                  "THURSDAY",
                  "FRIDAY",
                  "SATURDAY",
                  "SUNDAY",
                ],
                "startTimes": [
                  {
                    "hours": 8,
                  },
                ],
              },
            },
            "clusterType": "PRIMARY",
            "continuousBackupConfig": {
              "enabled": true,
              "recoveryWindowDays": 14,
            },
            "continuousBackupInfo": {
              "enabledTime": "2024-04-08T07:41:09.014516139Z",
              "encryptionInfo": {
                "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
              },
              "schedule": [
                "MONDAY",
                "TUESDAY",
                "WEDNESDAY",
                "THURSDAY",
                "FRIDAY",
                "SATURDAY",
                "SUNDAY",
              ],
            },
            "createTime": "2024-04-08T07:41:09.251865437Z",
            "databaseVersion": "POSTGRES_15",
            "encryptionInfo": {
              "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/asia-south1/clusters/test-cluster-03",
            "network": "projects/167984947943/global/networks/default",
            "networkConfig": {
              "network": "projects/167984947943/global/networks/default",
            },
            "reconciling": false,
            "state": "READY",
            "uid": "6a601812-cd9c-413f-b774-a7ad5fcfd8c0",
            "updateTime": "2024-04-11T10:52:14.682259204Z",
          },
        },
      ],
      "_type": "google_cloud_alloydb_cluster",
      "classification": "Database",
      "clusterName": "test-cluster-03",
      "clusterType": "PRIMARY",
      "createdOn": undefined,
      "databaseVersion": "POSTGRES_15",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/asia-south1/clusters/test-cluster-03",
      "encrypted": true,
      "hostname": "test-cluster-03",
      "initialUserName": undefined,
      "initialUserPassword": undefined,
      "kmsKeyName": undefined,
      "location": "asia-south1",
      "name": "projects/j1-gc-integration-dev-v3/locations/asia-south1/clusters/test-cluster-03",
      "networkConfigNetwork": "projects/167984947943/global/networks/default",
      "networkConfigallocatedIpRange": undefined,
      "reconciling": false,
      "state": "READY",
      "uid": "6a601812-cd9c-413f-b774-a7ad5fcfd8c0",
    },
    {
      "_class": [
        "Database",
        "DataStore",
        "Cluster",
      ],
      "_key": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/southamerica-west1/clusters/test-alloydb-cluster-2",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "automatedBackupPolicy": {
              "backupWindow": "3600s",
              "enabled": false,
              "location": "southamerica-west1",
              "timeBasedRetention": {
                "retentionPeriod": "1209600s",
              },
              "weeklySchedule": {
                "daysOfWeek": [
                  "MONDAY",
                  "TUESDAY",
                  "WEDNESDAY",
                  "THURSDAY",
                  "FRIDAY",
                  "SATURDAY",
                  "SUNDAY",
                ],
                "startTimes": [
                  {
                    "hours": 8,
                  },
                ],
              },
            },
            "clusterType": "PRIMARY",
            "continuousBackupConfig": {
              "enabled": true,
              "recoveryWindowDays": 14,
            },
            "continuousBackupInfo": {
              "enabledTime": "2024-04-08T06:04:05.775478728Z",
              "encryptionInfo": {
                "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
              },
              "schedule": [
                "MONDAY",
                "TUESDAY",
                "WEDNESDAY",
                "THURSDAY",
                "FRIDAY",
                "SATURDAY",
                "SUNDAY",
              ],
            },
            "createTime": "2024-04-08T06:04:06.022689568Z",
            "databaseVersion": "POSTGRES_14",
            "encryptionInfo": {
              "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/southamerica-west1/clusters/test-alloydb-cluster-2",
            "network": "projects/167984947943/global/networks/default",
            "networkConfig": {
              "network": "projects/167984947943/global/networks/default",
            },
            "reconciling": false,
            "state": "READY",
            "uid": "77ad0b2b-b129-491c-9954-a9526df1d684",
            "updateTime": "2024-04-11T10:39:04.810266479Z",
          },
        },
      ],
      "_type": "google_cloud_alloydb_cluster",
      "classification": "Database",
      "clusterName": "test-alloydb-cluster-2",
      "clusterType": "PRIMARY",
      "createdOn": undefined,
      "databaseVersion": "POSTGRES_14",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/southamerica-west1/clusters/test-alloydb-cluster-2",
      "encrypted": true,
      "hostname": "test-alloydb-cluster-2",
      "initialUserName": undefined,
      "initialUserPassword": undefined,
      "kmsKeyName": undefined,
      "location": "southamerica-west1",
      "name": "projects/j1-gc-integration-dev-v3/locations/southamerica-west1/clusters/test-alloydb-cluster-2",
      "networkConfigNetwork": "projects/167984947943/global/networks/default",
      "networkConfigallocatedIpRange": undefined,
      "reconciling": false,
      "state": "READY",
      "uid": "77ad0b2b-b129-491c-9954-a9526df1d684",
    },
    {
      "_class": [
        "Database",
        "DataStore",
        "Cluster",
      ],
      "_key": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-central1/clusters/test-alloydb-cluster-by-metron-team",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "automatedBackupPolicy": {
              "backupWindow": "3600s",
              "enabled": false,
              "location": "us-central1",
              "timeBasedRetention": {
                "retentionPeriod": "1209600s",
              },
              "weeklySchedule": {
                "daysOfWeek": [
                  "MONDAY",
                  "TUESDAY",
                  "WEDNESDAY",
                  "THURSDAY",
                  "FRIDAY",
                  "SATURDAY",
                  "SUNDAY",
                ],
                "startTimes": [
                  {
                    "hours": 8,
                  },
                ],
              },
            },
            "clusterType": "PRIMARY",
            "continuousBackupConfig": {
              "enabled": true,
              "recoveryWindowDays": 14,
            },
            "continuousBackupInfo": {
              "enabledTime": "2024-04-08T05:57:11.426010402Z",
              "encryptionInfo": {
                "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
              },
              "schedule": [
                "MONDAY",
                "TUESDAY",
                "WEDNESDAY",
                "THURSDAY",
                "FRIDAY",
                "SATURDAY",
                "SUNDAY",
              ],
            },
            "createTime": "2024-04-08T05:57:11.552941920Z",
            "databaseVersion": "POSTGRES_15",
            "encryptionInfo": {
              "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/clusters/test-alloydb-cluster-by-metron-team",
            "network": "projects/167984947943/global/networks/default",
            "networkConfig": {
              "allocatedIpRange": "test",
              "network": "projects/167984947943/global/networks/default",
            },
            "reconciling": false,
            "state": "READY",
            "uid": "5fb9ec08-bd7c-44c8-a231-1e72561d8649",
            "updateTime": "2024-04-11T11:54:35.404940450Z",
          },
        },
      ],
      "_type": "google_cloud_alloydb_cluster",
      "classification": "Database",
      "clusterName": "test-alloydb-cluster-by-metron-team",
      "clusterType": "PRIMARY",
      "createdOn": undefined,
      "databaseVersion": "POSTGRES_15",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-central1/clusters/test-alloydb-cluster-by-metron-team",
      "encrypted": true,
      "hostname": "test-alloydb-cluster-by-metron-team",
      "initialUserName": undefined,
      "initialUserPassword": undefined,
      "kmsKeyName": undefined,
      "location": "us-central1",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/clusters/test-alloydb-cluster-by-metron-team",
      "networkConfigNetwork": "projects/167984947943/global/networks/default",
      "networkConfigallocatedIpRange": "test",
      "reconciling": false,
      "state": "READY",
      "uid": "5fb9ec08-bd7c-44c8-a231-1e72561d8649",
    },
    {
      "_class": [
        "Database",
        "DataStore",
        "Cluster",
      ],
      "_key": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-01",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "automatedBackupPolicy": {
              "backupWindow": "3600s",
              "enabled": false,
              "location": "us-east4",
              "timeBasedRetention": {
                "retentionPeriod": "1209600s",
              },
              "weeklySchedule": {
                "daysOfWeek": [
                  "MONDAY",
                  "TUESDAY",
                  "WEDNESDAY",
                  "THURSDAY",
                  "FRIDAY",
                  "SATURDAY",
                  "SUNDAY",
                ],
                "startTimes": [
                  {
                    "hours": 8,
                  },
                ],
              },
            },
            "clusterType": "PRIMARY",
            "continuousBackupConfig": {
              "enabled": true,
              "recoveryWindowDays": 14,
            },
            "continuousBackupInfo": {
              "enabledTime": "2024-04-08T08:28:25.585092652Z",
              "encryptionInfo": {
                "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
              },
              "schedule": [
                "MONDAY",
                "TUESDAY",
                "WEDNESDAY",
                "THURSDAY",
                "FRIDAY",
                "SATURDAY",
                "SUNDAY",
              ],
            },
            "createTime": "2024-04-08T08:28:25.670796510Z",
            "databaseVersion": "POSTGRES_15",
            "encryptionInfo": {
              "encryptionType": "GOOGLE_DEFAULT_ENCRYPTION",
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-01",
            "network": "projects/167984947943/global/networks/test-network",
            "networkConfig": {
              "network": "projects/167984947943/global/networks/test-network",
            },
            "reconciling": false,
            "state": "READY",
            "uid": "5f16156c-e137-4533-9e88-e4ef13908c08",
            "updateTime": "2024-04-11T09:39:52.394640332Z",
          },
        },
      ],
      "_type": "google_cloud_alloydb_cluster",
      "classification": "Database",
      "clusterName": "cluster-01",
      "clusterType": "PRIMARY",
      "createdOn": undefined,
      "databaseVersion": "POSTGRES_15",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-01",
      "encrypted": true,
      "hostname": "cluster-01",
      "initialUserName": undefined,
      "initialUserPassword": undefined,
      "kmsKeyName": undefined,
      "location": "us-east4",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-01",
      "networkConfigNetwork": "projects/167984947943/global/networks/test-network",
      "networkConfigallocatedIpRange": undefined,
      "reconciling": false,
      "state": "READY",
      "uid": "5f16156c-e137-4533-9e88-e4ef13908c08",
    },
    {
      "_class": [
        "Database",
        "DataStore",
        "Cluster",
      ],
      "_key": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-05",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "automatedBackupPolicy": {
              "backupWindow": "3600s",
              "enabled": false,
              "location": "us-east4",
              "timeBasedRetention": {
                "retentionPeriod": "1209600s",
              },
              "weeklySchedule": {
                "daysOfWeek": [
                  "MONDAY",
                  "TUESDAY",
                  "WEDNESDAY",
                  "THURSDAY",
                  "FRIDAY",
                  "SATURDAY",
                  "SUNDAY",
                ],
                "startTimes": [
                  {
                    "hours": 8,
                  },
                ],
              },
            },
            "clusterType": "PRIMARY",
            "continuousBackupConfig": {
              "enabled": true,
              "encryptionConfig": {
                "kmsKeyName": "projects/j1-gc-integration-dev-v3/locations/us-east4/keyRings/test-Keyring-by-metron/cryptoKeys/High",
              },
              "recoveryWindowDays": 14,
            },
            "continuousBackupInfo": {
              "enabledTime": "2024-04-10T11:23:01.036554878Z",
              "encryptionInfo": {
                "encryptionType": "CUSTOMER_MANAGED_ENCRYPTION",
                "kmsKeyVersions": [
                  "projects/j1-gc-integration-dev-v3/locations/us-east4/keyRings/test-Keyring-by-metron/cryptoKeys/High/cryptoKeyVersions/1",
                ],
              },
              "schedule": [
                "MONDAY",
                "TUESDAY",
                "WEDNESDAY",
                "THURSDAY",
                "FRIDAY",
                "SATURDAY",
                "SUNDAY",
              ],
            },
            "createTime": "2024-04-10T11:23:01.251059714Z",
            "databaseVersion": "POSTGRES_15",
            "encryptionConfig": {
              "kmsKeyName": "projects/j1-gc-integration-dev-v3/locations/us-east4/keyRings/test-Keyring-by-metron/cryptoKeys/High",
            },
            "encryptionInfo": {
              "encryptionType": "CUSTOMER_MANAGED_ENCRYPTION",
              "kmsKeyVersions": [
                "projects/j1-gc-integration-dev-v3/locations/us-east4/keyRings/test-Keyring-by-metron/cryptoKeys/High/cryptoKeyVersions/1",
              ],
            },
            "name": "projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-05",
            "network": "projects/167984947943/global/networks/test-network",
            "networkConfig": {
              "network": "projects/167984947943/global/networks/test-network",
            },
            "reconciling": false,
            "state": "READY",
            "uid": "2d2051af-2261-4a36-9701-080bf296ee2b",
            "updateTime": "2024-04-11T09:39:52.776820230Z",
          },
        },
      ],
      "_type": "google_cloud_alloydb_cluster",
      "classification": "Database",
      "clusterName": "cluster-05",
      "clusterType": "PRIMARY",
      "createdOn": undefined,
      "databaseVersion": "POSTGRES_15",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-05",
      "encrypted": true,
      "hostname": "cluster-05",
      "initialUserName": undefined,
      "initialUserPassword": undefined,
      "kmsKeyName": "projects/j1-gc-integration-dev-v3/locations/us-east4/keyRings/test-Keyring-by-metron/cryptoKeys/High",
      "location": "us-east4",
      "name": "projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-05",
      "networkConfigNetwork": "projects/167984947943/global/networks/test-network",
      "networkConfigallocatedIpRange": undefined,
      "reconciling": false,
      "state": "READY",
      "uid": "2d2051af-2261-4a36-9701-080bf296ee2b",
    },
    {
      "_class": [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "W/"8f56735f6f3d890d"",
            "labels": {
              "firebase": "enabled",
            },
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2024-02-27T16:17:58.322524Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "tag.firebase": "enabled",
      "updatedOn": 1709050678322,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/asia-south1/clusters/test-cluster-03",
      "_toEntityKey": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/asia-south1/clusters/test-cluster-03",
      "_type": "google_cloud_project_has_alloydb_cluster",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/southamerica-west1/clusters/test-alloydb-cluster-2",
      "_toEntityKey": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/southamerica-west1/clusters/test-alloydb-cluster-2",
      "_type": "google_cloud_project_has_alloydb_cluster",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-central1/clusters/test-alloydb-cluster-by-metron-team",
      "_toEntityKey": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-central1/clusters/test-alloydb-cluster-by-metron-team",
      "_type": "google_cloud_project_has_alloydb_cluster",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-01",
      "_toEntityKey": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-01",
      "_type": "google_cloud_project_has_alloydb_cluster",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-05",
      "_toEntityKey": "google_cloud_alloydb_cluster:projects/j1-gc-integration-dev-v3/locations/us-east4/clusters/cluster-05",
      "_type": "google_cloud_project_has_alloydb_cluster",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": [
    "google_cloud_alloydb_cluster",
    "google_cloud_project",
    "google_cloud_project_has_alloydb_cluster",
  ],
  "numCollectedEntities": 6,
  "numCollectedRelationships": 5,
}
`;
